### ISSUE TITLE
 What stands behind this idea ? 

### ISSUE THREAD

**Issue by RRadziejewski** [psdlt/roboape#1]
**The idea is that given enough time, bot will build up enough good tokens to generate constant revenue. But to get there you will need to first burn through those rug pulls.**

What stands behind this idea ? Any tests? How much initial value in BNB an user need to have in order to become profitable with this. I know there is no golden rule for sure, i just ask for rough estimation.

**Comment by psdlt** [comment]
@RRadziejewski you're not wrong there - there is no golden rule, as there are a lot of moving parts:
- A number of new LPs created. [See PancakeSwap factory here](https://bscscan.com/txsInternal?a=0xcA143Ce32Fe78f1f7019d7d551a6402fC5350c73&p=1). In the last 24 hours there were ~4500 LPs created. Multiply that by your buy in (default is 0.002 BNB) and you get a rough estimate (~9 BNB) if you want to participate in all of them and you don't take any profits. Of course, not all of those LPs are between <token> and BNB, but a vast majority is.
- There are two broad categories of tokens - those that have their contract source published ("verified") on BSScan and those who don't. You can choose to skip "unverified" tokens and that will cut your expenditure by more than half.
- Analysis of token source code for "bad words" will further eliminate quite a few LPs.
- Keep in mind that _some_ of the buy in will yield profits within first few hours.

Considering all that, at the very least you should expect to spend 3-5 BNB within the first 24 hours. With time it _should_ get better.

As for tests, I have done limited testing to make sure the bot works. Don't have enough "play money" to fully verify the theory. But what you can do is you can analyse other bots that do exact same thing. Just open any LP (i.e. [this one](https://bscscan.com/address/0x212306e238598b58d3531c678cbf3b2d3b45b488#tokentxns). nothing special about it, picked it randomly from the list linked above), go the first page of BEP20 token txns and you will find 4-8 bots each buying 0.002, some buying more. Open their accounts and "do your own research" 🙂 BSScan provides a very nice "Analytics" overview of certain aspects (balances, transfers, txn count) of an active account.

I hope this info helps and as always - not a financial advice! 😁

**Comment by krim6ix** [comment]
You have promoted this on bscscan so much this is literally in every bot controlled account's comment section . []([]([.]([url](https://bscscan.com/address/0x5e3773713f0a20806168d633652b905ee7253f83#comments)))) Interesting part is that many people are replying to your comment that this script transferd there bnb and claiming this to be a scam. You haven't replied to any of them . As a dev who made this i think you should atleast have tried to know what went wrong with them if you are asking people to use your script. I still don't know who true their claim is.

**Comment by psdlt** [comment]
@krim6ix let me start by saying that those comments are false :)
I'm not entirely sure why people make those claims (about lost BNB), but my guess would be - they don't want competition. If I wanted to steal anyones crypto - I would be distributing a compiled binary and I would say that "this bot will 100x your investment within one hour"... But I'm releasing the source code for anyone to read, and that's exactly what you should do before using any similar bot - read the source! Make sure there is nothing fishy going on there! Use your own judgement. And in _this_ bot's case, there isn't that much source there, a few files, so it shouldn't take long to check whether it's good or not.
Anyway, thank you for your concern :) Smart apes will have no trouble seeing through the FUD these other people are spreading ;)

**Comment by ARebek** [comment]
Well, I've a guess why those commenters are angry with you. I also believe those comments are false, but using the bot could make you loose a lot of BNBs.

I've tried it, used it, it earned a lot BNBs but on the other hand I needed to fuel it with more BNBs than it earned. I haven't count it yet, but the purchases and the fees ate more than 10 BNBs already. What am I doing wrong? I used almost the default config and I didn't let the bot to buy any unverified tokens, either. After 10 days I can report that using it (at least in that way I used) is just not profitable. Rather a money pit. If you can suggest something please do it.

The second thing. Just wondering.. If the bot is profitable for you, why whould you share it's code to make competion for yourself ? It doesn't seem logical to me. Except if it already made you richy rich and you doesn't care anymore :)

**Comment by psdlt** [comment]
@ARebek first of all, thank you for you comment :)

As stated earlier, I only did limited testing, but during that testing I found that setting `AUTOSELL_MIN_PROFIT` to lower values seems to yield better results. You could also try to play around with the value (https://github.com/psdlt/roboape/blob/main/src/Services/ProfitLossManager.ts#L28) which sets how long bot should wait before checking any given token for profits (currently hardcoded to `30` mintues).

Overall, this bot is more of a gamble than a solid investment strategy. It depends a lot on overall network activity and token flows (i.e. how many good tokens are launched vs how many rug pullers are out there).

Why would I release this bot? A couple of reason. One is that crypto for me is a hobby, I don't expect to get rich from this. Second is that I simply like writing code and this bot was a fun little exercise :) And on top of all that - I just want to give something back to the community.

Hope this helps!

**Comment by TJCares1966** [comment]
I keep getting this error:

[DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.

What did I do wrong? 

Thanks for the help

**Comment by psdlt** [comment]
@TJCares1966 you didn't do anything wrong. It's a warning (not an error) thrown by one of the libraries used by this bot. I'm not exactly sure which one library that is, but I think that might be `web3js`. Hopefully library vendor will fix this issue in future releases.

Either way, `roboape` works just fine even with this warning :)

**Comment by TJCares1966** [comment]
No it doesn't, stops it cold. would you like to see the whole error?

**Comment by psdlt** [comment]
@TJCares1966 the error/warning you quoted is definitely not an issue. If there is additional output, please share it, but please open a separate issue not to pollute this one :)

**Comment by ARebek** [comment]
@psdlt You are welcome.

Thank you for your ideas. I will give them a last last try. I am starting to believe this summer-time is not the best for this business. Too few good quality tokens and too much rug pulls. What do you think ?

A little different topic, just for fun. Can you imagine there is a bot out there which is copying my bot's every movement, BUT it does it in realtime, moreover, it's detecting my gas and doing the exactly same operation but with higher gas. I mean if I start to buy / sell a specific token which also exists in the copier bot's portfolio, it starts to buy / sell the same thing in the same second but with a little bit higher gas. Of course making my sellings / buying more expensive / less effective.  I also copying approvements is, like my bot's mirror ghost bot.

**Comment by psdlt** [comment]
@ARebek yea, summer seems to be slow, not that much network activity as there was 2-3 months ago.

> it's detecting my gas and doing the exactly same operation but with higher gas

that's a technique called `front running`, and unfortunately it's a common thing.

**Comment by gensx-x1** [comment]
> Well, I've a guess why those commenters are angry with you. I also believe those comments are false, but using the bot could make you loose a lot of BNBs.
> 
> I've tried it, used it, it earned a lot BNBs but on the other hand I needed to fuel it with more BNBs than it earned. I haven't count it yet, but the purchases and the fees ate more than 10 BNBs already. What am I doing wrong? I used almost the default config and I didn't let the bot to buy any unverified tokens, either. After 10 days I can report that using it (at least in that way I used) is just not profitable. Rather a money pit. If you can suggest something please do it.
> 
> The second thing. Just wondering.. If the bot is profitable for you, why whould you share it's code to make competion for yourself ? It doesn't seem logical to me. Except if it already made you richy rich and you doesn't care anymore :)

it ate 10bnb because this script relies on bsc scan to check for bad words , however i researched one of wallet and on avarage it buys around 2k tokens and pancake factory on average produce 7k lp's daily.
@psdlt your work is awsome! im making python version for personal use and your script help me a lot to develop it, for future consider implementing EVM decompiler to decompile byte code and look for bad words and/or bad functions. dm me if you want link to my wallet on bscscan. cheers

**Comment by ARebek** [comment]
Thanks 4 your answer! There is always something new to learn.. Thank you! I guess the answer, but is there anything you can prevent this front running thing? It already stopped doing it but was very very annoying.

This is something similar what I've just experienced some days ago. There was a coin, I put 9 BNB in it and some days later I decided to take it out. There was no transaction, it became a slow, calm token. I hit the Swap button on poocoin then - surprise, after some seconds the page said my transaction was reverted by EVM. Okay. I hit the button again. Sold.

Then a bit later I've checked my selling and I saw on bscscan in the same second when I sold my tokens there was an another transaction, too. An another account just bought almost for the same price (I mean +-0.1-0.2 BNB) the same token. I did not feel I got less BNBs for my tokens but it was strange. (Or perhaps I just didn't notice that I was drained by some bucks, but I really dont think so) But then what was the other guy/bot's purpose..? There is only one sure thing. Fortunately no more strange stories :)

**Comment by psdlt** [comment]
@ARebek front-running is sort-of kind-of mitigable, but it's definitely outside the scope of this project/bot and it also has it's own limitations.

As for your experience with that 9BNB txn: difficult to say without seeing txns, but my guess would be arbitrage.

**Comment by ARebek** [comment]
@psdlt : Yeah, this kind of unfair front running behaviour is far from this bot's goal I know. So sorry for being offtopic a bit here.

When your bot is trying to sell something more valuable I'm sure it's also a victim of these front runner bots as well. I already saw more times the error message Reverted by EVM on the screen when your bot wanted to sell something. Not all the time but sometimes it happens. So this thingy is much more annoying than I thought first. I just made a test just to be sure my sellings is under manipulation. I just wanted to sell a little quantity from my coins let's say for 1 BNB. I clicked swap button then before I hit the final metamask  Accept button I modified the gwei from the default 5 to 21, then I hit the accept. My transaction (of course) was reverted by EVM, and nothing happened. At the same time a selling appeared in the trx list what I opened and not a big surprise, an another account sold the near same quality of tokens - and what do you think what was the transaction's gwei value? Yes, a bit more than mine, it was 22. I think this is more than a clear proof. What I don't know are they spotting me (I think not) or just all the bigger transactions' of some coins. Anyways it's a very nasty way to steal money from people. If you have any idea against them please tell me, thanks in advance. You said it has it's own limitations, to tell the truth this was what made me curious. :)

**Comment by ARebek** [comment]
Hey,

I didn't know there is no private message on github. Or I'm just too dumb to find it, whatever.

Please check your accounts (especially this one: 0x5e3773...........e7253f83) , because you will loose a LOT of BNBs on JEDSTAR! What I suppose you really don't want to. :)

https://poocoin.app/tokens/0x058a7af19bdb63411d0a84e79e3312610d7fa90c

Sorry for being so pubic here but you can delete this message if it contains any sensitive data, but may be not.

Cheers.
Rebek

**Comment by ARebek** [comment]
Hello @psdlt !

I am back again. But I hope I am not annoying. :)) So.. Can you examine a little this address, please?
https://bscscan.com/address/0x5e3773713f0A20806168D633652B905ee7253f83
It has 225K transactions so far so must be a bot for sure.

It looks in the beginning (76 days ago..) it started to buy new coins with a bot just like yours. But what I noticed this address started to do front running attacks, too - at least some time ago. [ I just learned here what front running is if you remember.. ] 

So this was the original transaction with gwei 5: _[ failed and reverted ]_
https://bscscan.com/tx/0xe85caa1867949fc4803841822e51eabb7aa90890975694d69fbda33119e84802 

And the front runner's transaction with 5.00002 gwei:  [succed]
https://bscscan.com/tx/0x8d30e0f24d007726ac3219fa67f67187d15c08e98d4eed6032ef8b4868bdec30

What do you think?

Regards,
Rebek

**Comment by psdlt** [comment]
@ARebek that certainly looks like a more advanced bot :)

**Comment by ARebek** [comment]
yepp.. for sure..
U mean a combined bot which can purchase new coins and also frontrun?
We can call it roboape 2.0 :)

Another thought: once you said operating with lower autosell setting the bot works better as you experienced. Is this still true? I mean it may look more profitable at the first sight, but it will also sell the really superb coins before they can reach their real potential.

**Comment by gensx-x1** [comment]
just a small tip , you need to run your own full node on bsc to do any frontrun action , i've tried it already. running your own node is pretty expensive. without using public node your transaction will be always 2 blocks too late no matter what language you'll use to write script

**Comment by ARebek** [comment]
thanks.
